### ISSUE TITLE
[WIP] Lazy NBT approach

### ISSUE THREAD

**Issue by toqueteos** [PistonDevelopers/hematite_server#67]
# THIS IS A WORK IN PROGRESS

_Reference branch: https://github.com/toqueteos/hematite_server/tree/chunk-loader_

Hello guys,

Some of you know I've been working on reading world info from region/`.mca` files. It turned out to be **awfully slow** due to our current NBT enc/decoder doing waaay too many allocations.

> **Quick intro:** Region files are composed of a 8KB header (2 groups of 1024 integers) and an arbitrary number `N` of 4KB blocks storing chunk data. Each of these blocks is compressed with zlib. More info: http://minecraft.gamepedia.com/Region_file_format.

Initially, I had a program which read all region files inside `.minecraft/saves` but it was taking BOATLOADS of time to read all those files (I was expecting it to be faster) so I reduced my test case to just one single region file with 712 chunks (max is 1024), ran a naive benchmark and found that it was taking **27 seconds**. That's **38ms per chunk**.

So I began exploring ways of improving performance (@eddyb suffered my questions and suggested a lazy approach). Results are in https://github.com/toqueteos/rust-nbt and for the same exact region file it went down to **700ms**, that's **1ms per chunk**.

38x perf increase.

I tried with my initial approach of reading ALL region files and it also worked blazingly fast.
## So... how does it (decoding) work?

_(I only have the decoder part working, it seems this doesn't work for encoding, so we may end up having `NbtReader` and `NbtWriter`)_
- Read the Nbt file/chunk/blob/whatever into memory, all of it.
- Read every `NbtValue`'s header, save its position and length; skipping value decoding unless it's a `List` or `Compound`.
## Notes
- Not finished! Only decoding working.
- Probably O(n) cost for indexing.
- We can't access items by their name right now.
- NbtValue is gone (might change). The equivalent is called Entry.
- Not as straightforward as our current recursive impl.
- `Iterator` seems quite easy to do.
- Currently working on `fn get<T: FromNbt>(name: &str) -> T`. This requires having some sort of string-index with some a well defined naming scheme so we can get values n-levels deep inside the Nbt tree (example: `a:b:c` or `a/b/c`).

/cc @atheriel @fenhl @eddyb 
## Performance numbers

Using `cargo run --release`:

```
read_mca_file ".minecraft\saves\World1\region\r.-1.0.mca"   88/1024 (8.59)
elapsed (OLD / NEW): 254ms / 35ms
read_mca_file ".minecraft\saves\World1\region\r.-1.1.mca"    2/1024 (0.20)
elapsed (OLD / NEW): 11ms / 5ms
read_mca_file ".minecraft\saves\World1\region\r.0.0.mca"   712/1024 (69.53)
elapsed (OLD / NEW): 2150ms / 277ms
read_mca_file ".minecraft\saves\World1\region\r.0.1.mca"    24/1024 (2.34)
elapsed (OLD / NEW): 81ms / 12ms
read_mca_file ".minecraft\saves\World2\region\r.-1.0.mca"  714/1024 (69.73)
elapsed (OLD / NEW): 2452ms / 351ms
read_mca_file ".minecraft\saves\World3\region\r.-1.-1.mca" 641/1024 (62.60)
elapsed (OLD / NEW): 1436ms / 218ms
read_mca_file ".minecraft\saves\World3\region\r.-1.-2.mca" 422/1024 (41.21)
elapsed (OLD / NEW): 949ms / 141ms
read_mca_file ".minecraft\saves\World3\region\r.-1.0.mca"  105/1024 (10.25)
elapsed (OLD / NEW): 234ms / 41ms
read_mca_file ".minecraft\saves\World3\region\r.-2.-1.mca"  45/1024 (4.39)
elapsed (OLD / NEW): 102ms / 23ms
read_mca_file ".minecraft\saves\World3\region\r.-2.-2.mca"  77/1024 (7.52)
elapsed (OLD / NEW): 175ms / 29ms
read_mca_file ".minecraft\saves\World3\region\r.0.-1.mca"  124/1024 (12.11)
elapsed (OLD / NEW): 280ms / 44ms
read_mca_file ".minecraft\saves\World3\region\r.0.-2.mca"   51/1024 (4.98)
elapsed (OLD / NEW): 117ms / 20ms
read_mca_file ".minecraft\saves\World3\region\r.0.0.mca"    11/1024 (1.07)
elapsed (OLD / NEW): 29ms / 7ms
```

**Comment by atheriel** [comment]
I think it's generally a good idea to optimize for performance, but I have a few thoughts:
- Does this actually make decoding the whole structure faster, or just getting it into memory? I think those are probably different goals...
- Does this mean we have to wait for the TCP stream to finish before we start decoding? If so, what's the point of the `from_reader` approach?
- Is the idea that the laziness is tied to indexing, so that it won't decode the `blob['playerHealth']` section of memory unless you access it?

There was a discussion of the nasty, nasty hacks involved in getting `html` parsing working with no allocations as a part of the `html5ever` project on reddit a few days ago.

**Comment by toqueteos** [comment]
@atheriel Those are all good points!
- All it does is making value decoding lazy; instead of allocating a `String` and some variable size `T` for every Nbt entry so.. yes, I'd say it does decode the whole structure faster. The initial idea was not to copy data we already have.
- No matter what approach we use need to read it all, do we?. What's wrong with `from_reader`? This works for in-memory streams, files and network streams.
- That's right, unless it's a Compound or List; if it is it will scan its children recursively annotating them all.

I may have forgot to write that down but this is intended for reading region (`.mca`) files from disk.
Trying to emulate a server start is how I found our current approach (for Nbt decoding) wasn't working _"fast enough"_.

If we have fast Nbt, server starts fast and we can also send chunks to net fast. Right now sending chunks through network requires reading from disk to our in-memory repr and encoding back to a byte stream. We could _mostly_ sidestep this with this new approach.

I'm biased because I made it but it shouldn't be hacky at all. It just reads ID, stores index position as start, skips the name and value (unless it's a List or Compound) and stores index position as end for every entry.
